Lib/enum.py

```diff
@@ -728,6 +728,10 @@ def __contains__(self, other):
                     type(other).__qualname__, self.__class__.__qualname__))
         return other._value_ & self._value_ == other._value_
 
+    def __iter__(self):
+        members, extra_flags = _decompose(self.__class__, self.value)
+        return (m for m in members if m._value_ != 0)
+
```

Comment:
The `_decompose` function is heavyweight (fairly complex and includes sorting).    I wouldn't expect this kind of overhead when iterating flag bits.

I can think of two efficient ways:

1. compare to all members of the Flag class, excluding 0 and non-power-of-2:
```python3
return (v for v in self.__class__ if v.value != 0 and (v.value & (v.value-1) == 0) and v in self)
```

2. compare to `~self.__class__(0)` with a helper to iterate bits
```python3
def _bits(n):
    while n:
        b = n & (~n+1)
        yield b
        n ^= b

...
return (v for v in map(self.__class__, _bits(~self.__class__(0))) and v in self)
```

really the set of actual bits should be stored in a class attribute, so it doesn't need to be computed

Reply:
proposal (2) is 2x faster than the implementation in this PR, for an enum with only about 6 values...

```python3
>>> from enum import Flag, auto
>>> class Foo(Flag):
...     NONE = 0
...     A = auto()
...     B = auto()
...     C = auto()
...     C2 = C
...     AB = A | B
...     D = auto()
...     E = auto()
...     F = auto()
...
>>> x = Foo.A | Foo.B | Foo.D | Foo.F
>>> import timeit
>>> timeit.timeit('list(x)', globals=locals(), number=10000)
0.6332780510000013
```
proposal (2):
```python3
>>> timeit.timeit('list(x)', globals=locals(), number=10000)
0.36078684099999236
```

Reply:
I have become increasingly unhappy with the the multiple calls to `_decompose`.  Your ideas (and others, too) are welcome.

Reply:
> I have become increasingly unhappy with the the multiple calls to _decompose.  Your ideas (and others, too) are welcome.

I do have ideas on it, and intended to open a separate issue for refactoring `_decompose()` and make a PR replacing it with something simpler, faster, and less buggy.  (E.g. use the `_bits()` iterator; and I noticed that all callers of it either use the `members` or `extra_flags` output, never both;  and the computation of `extra_flags` would be trivial if we had an `__all__` attribute on the class.)

Regarding bugs, when it's used by `repr()` or `str()`, I don't think it's intended that a lone compound value is represented as-is, while a compound plus other value is represented as a redundant "compound plus underlying bits":

```python3
>>> class Foo(Flag):
...     A = auto()
...     B = auto()
...     C = auto()
...     BC = B | C
>>> repr(Foo.BC)
'<Foo.BC: 6>'
>>> repr(Foo.A | Foo.BC)
'<Foo.BC|C|B|A: 7>'
```